docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Third parties (e.g., open shift) may have their own implementation.
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
```

Comment:
I would go even further and state "rest client is generically usable by any client exposing rest like semantics.  It provides helpers that benefit the Kube like API conventions, but must not mandate them.  It provides a higher level http interface that abstracts transport, wire serialization, retry logic, and error handling.  Kube like constructs that deviate from standard http should be bypassable.  *every* non trivial call made to a remote restful API from Kube code should go through a rest client.